frontend/server/app.ts

```diff
@@ -0,0 +1,161 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import * as path from 'path';
+import * as express from 'express';
+import { Application, static as StaticHandler } from 'express';
+import * as proxy from 'http-proxy-middleware';
+
+import { IConfigs } from './configs';
+import { getAddress } from './utils';
+import { getBuildMetadata, getHealthzEndpoint, getHealthzHandler } from './handlers/healthz';
+import { getArtifactsHandler } from './handlers/artifacts';
+import { getCreateTensorboardHandler, getGetTensorboardHandler } from './handlers/tensorboard';
+import { getPodLogsHandler } from './handlers/pod-logs';
+import { clusterNameHandler, projectIdHandler } from './handlers/core';
+import { getAllowCustomVisualizationsHandler } from './handlers/vis';
+import { getIndexHTMLHandler } from './handlers/index-html';
+
+import proxyMiddleware from './proxy-middleware';
+import { Server } from 'http';
+
+function getRegisterHandler(app: Application, basePath: string) {
+  return (
+    func: (name: string, handler: express.Handler) => express.Application,
+    route: string,
+    handler: express.Handler,
+  ) => {
+    func.call(app, route, handler);
+    return func.call(app, `${basePath}${route}`, handler);
+  };
+}
+
+export class UIServer {
+  app: Application;
+  httpServer?: Server;
+
+  constructor(public readonly options: IConfigs) {
+    this.app = createUIServer(options);
+  }
+
+  start(port?: number | string) {
+    port = port || this.options.server.port;
+    this.httpServer = this.app.listen(port, () => {
```

Comment:
nit: also throw an error if http server is already initialized?

Reply:
fixed.